src/interpreter/global.rs

```diff
@@ -0,0 +1,298 @@
+
+use error::EvalResult;
+use memory::{Memory, Pointer};
+
+use std::collections::HashMap;
+
+use interpreter::function::FnEvalContext;
+use interpreter::{
+    ConstantId,
+    CachedMir,
+};
+
+pub(super) struct GlobalEvalContext<'a, 'tcx: 'a> {
+    /// The results of the type checker, from rustc.
+    pub tcx: TyCtxt<'a, 'tcx, 'tcx>,
+
+    /// A mapping from NodeIds to Mir, from rustc. Only contains MIR for crate-local items.
+    mir_map: &'a MirMap<'tcx>,
+
+    /// A local cache from DefIds to Mir for non-crate-local items.
+    mir_cache: RefCell<DefIdMap<Rc<mir::Mir<'tcx>>>>,
+
+    /// The virtual memory system.
+    pub memory: Memory,
+
+    /// Precomputed statics, constants and promoteds
+    pub statics: HashMap<ConstantId<'tcx>, Pointer>,
+
+    /// Monomorphising types is slow and done often
+    monomorphizations: RefCell<HashMap<(Ty<'tcx>, &'tcx Substs<'tcx>), &'tcx Layout>>,
+}
+
+
+impl<'a, 'tcx> GlobalEvalContext<'a, 'tcx> {
+    pub fn new(tcx: TyCtxt<'a, 'tcx, 'tcx>, mir_map: &'a MirMap<'tcx>) -> Self {
+        GlobalEvalContext {
+            tcx: tcx,
+            mir_map: mir_map,
+            mir_cache: RefCell::new(DefIdMap()),
+            memory: Memory::new(tcx.sess
+                                   .target
+                                   .uint_type
+                                   .bit_width()
+                                   .expect("Session::target::uint_type was usize")/8),
+            statics: HashMap::new(),
+            monomorphizations: RefCell::new(HashMap::new()),
+        }
+    }
+
+    pub fn call(&mut self, mir: &mir::Mir<'tcx>, def_id: DefId) -> EvalResult<Option<Pointer>> {
+        let substs = self.tcx.mk_substs(subst::Substs::empty());
+        let return_ptr = self.alloc_ret_ptr(mir.return_ty, substs);
+
+        let mut nested_fecx = FnEvalContext::new(self);
+
+        nested_fecx.push_stack_frame(def_id, mir.span, CachedMir::Ref(mir), substs, None);
+
+        nested_fecx.frame_mut().return_ptr = return_ptr;
+
+        nested_fecx.run()?;
+        Ok(return_ptr)
+    }
+
+    pub fn alloc_ret_ptr(&mut self, ty: ty::FnOutput<'tcx>, substs: &'tcx Substs<'tcx>) -> Option<Pointer> {
+        match ty {
+            ty::FnConverging(ty) => {
+                let size = self.type_size(ty, substs);
+                Some(self.memory.allocate(size))
+            }
+            ty::FnDiverging => None,
+        }
+    }
+
+    pub fn monomorphize(&self, ty: Ty<'tcx>, substs: &'tcx Substs<'tcx>) -> Ty<'tcx> {
+        let substituted = ty.subst(self.tcx, substs);
+        self.tcx.normalize_associated_type(&substituted)
+    }
+
+    pub fn type_size(&self, ty: Ty<'tcx>, substs: &'tcx Substs<'tcx>) -> usize {
+        self.type_layout(ty, substs).size(&self.tcx.data_layout).bytes() as usize
+    }
+
+    pub fn type_layout(&self, ty: Ty<'tcx>, substs: &'tcx Substs<'tcx>) -> &'tcx Layout {
+        if let Some(layout) = self.monomorphizations.borrow().get(&(ty, substs)) {
```

Comment:
I undid all the caching, because looking at the traces it's not really clear what's going on. Often times the "same" (textual output is the same) type + subst combo get hashed differently... I don't know enough about this do to it properly